zipline/utils/calendars/calendar_helpers.py

```diff
@@ -0,0 +1,213 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pandas as pd
+import numpy as np
+import bisect
+
+from zipline.errors import NoFurtherDataError
+
+
+def normalize_date(date):
+    date = pd.Timestamp(date, tz='UTC')
+    return pd.tseries.tools.normalize_date(date)
+
+
+def delta_from_time(t):
+    """
+    Convert a datetime.time into a timedelta.
+    """
+    return pd.Timedelta(
+        hours=t.hour,
+        minutes=t.minute,
+        seconds=t.second,
+    )
+
+
+def _get_index(dt, all_trading_days):
+    """
+    Return the index of the given @dt, or the index of the preceding
+    trading day if the given dt is not in the trading calendar.
+    """
+    ndt = normalize_date(dt)
+    if ndt in all_trading_days:
+        return all_trading_days.searchsorted(ndt)
+    else:
+        return all_trading_days.searchsorted(ndt) - 1
+
+# The following methods are intended to be inserted in both the
+# ExchangeCalendar and TradingSchedule classes with partial hooks to those
+# class' methods. These methods live in the helpers module to avoid code
+# duplication.
+
+
+def next_scheduled_day(date, last_trading_day, is_scheduled_day_hook):
+    dt = normalize_date(date)
+    delta = pd.Timedelta(days=1)
+
+    while dt <= last_trading_day:
+        dt += delta
+        if is_scheduled_day_hook(dt):
+            return dt
+    return None
+
+
+def previous_scheduled_day(date, first_trading_day, is_scheduled_day_hook):
+    dt = normalize_date(date)
+    delta = pd.Timedelta(days=-1)
+
+    while first_trading_day < dt:
+        dt += delta
+        if is_scheduled_day_hook(dt):
+            return dt
+    return None
+
+
+def next_open_and_close(date, open_and_close_hook,
+                        next_scheduled_day_hook):
+    return open_and_close_hook(next_scheduled_day_hook(date))
+
+
+def previous_open_and_close(date, open_and_close_hook,
+                            previous_scheduled_day_hook):
+    return open_and_close_hook(previous_scheduled_day_hook(date))
+
+
+def scheduled_day_distance(first_date, second_date, all_days):
+    first_date = normalize_date(first_date)
+    second_date = normalize_date(second_date)
+
+    i = bisect.bisect_left(all_days, first_date)
+    if i == len(all_days):  # nothing found
+        return None
+    j = bisect.bisect_left(all_days, second_date)
+    if j == len(all_days):
+        return None
+    distance = j - 1
+    assert distance >= 0
+    return distance
+
+
+def minutes_for_day(day, open_and_close_hook):
+    start, end = open_and_close_hook(day)
+    return pd.date_range(start, end, freq='T')
+
+
+def days_in_range(start, end, all_days):
```

Comment:
Was this the previously-existing implementation?  This is doing four linear scans of a large array, when it should be doing two binary searches.  The right implementation of this is:

```
return all_days[all_days.slice_indexer(start_date, end_date)]
```